tests/tools/plugins/test_msg_comp.cc

```diff
@@ -0,0 +1,263 @@
+ */
+
+/*
+Regression test code for TS CPP API header HttpMsgComp.h.  The code assumes there will only be one active transaction at a time.
+*/
+
+#include <fstream>
+#include <cstdlib>
+#include <string_view>
+
+#include <ts/ts.h>
+
+#include <tscpp/api/HttpMsgComp.h>
+
+// TSReleaseAssert() doesn't seem to produce any logging output for a debug build, so do both kinds of assert.
+//
+#if defined(__OPTIMIZE__)
+
+#define ALWAYS_ASSERT(EXPR) \
+  {                         \
+    TSReleaseAssert(EXPR);  \
+  }
+
+#else
+
+#define ALWAYS_ASSERT(EXPR) \
+  {                         \
+    TSAssert(EXPR);         \
+  }
+
+#endif
+
+namespace
+{
+#define PINAME "msg_comp"
+char PIName[] = PINAME;
+
+// NOTE:  It's important to flush this after writing so that a gold test using this plugin can examine the log before TS
+// terminates.
+//
+std::fstream logFile;
+
+using atscppapi::txnRemapFromUrlStringGet;
+using atscppapi::txnRemapToUrlStringGet;
+using atscppapi::txnEffectiveUrlStringGet;
+using atscppapi::MsgBase;
+using atscppapi::MimeField;
+using atscppapi::TxnClientReq;
+using atscppapi::TxnClientResp;
+using atscppapi::TxnServerReq;
+using atscppapi::TxnServerResp;
+
+void
+dumpMimeField(const MimeField &fld)
+{
+  logFile << fld.nameGet() << ": ";
+  int nVals = fld.valuesCount();
+
+  if (nVals > 0) {
+    logFile << fld.valGet(0);
+
+    for (int i = 1; i < nVals; ++i) {
+      logFile << ", " << fld.valGet(i);
+    }
+  }
+  logFile << std::endl;
+}
+
+bool
+sameMimeField(const MimeField &f1, const MimeField &f2)
+{
+  if (&f1.msg() != &f2.msg()) {
+    return false;
+  }
+
+#if 0
+
+  if (&f1.loc() != &f2.loc()) {
+    return false;
+  }
+
+#else
+
+  // Presumably as some sort of homage to Satan, trafficserver seems to make duplicate copies of the same Mime header
+  // within a message.
+  //
+  int nVals = f1.valuesCount();
+  if (nVals != f2.valuesCount()) {
+    return false;
+  }
+  for (int i = 0; i < nVals; ++i) {
+    if (f2.valGet(i) != f2.valGet(i)) {
+      return false;
+    }
+  }
+
+#endif
+
+  return true;
+}
+
+void
+dumpMsg(MsgBase &msg, std::string_view msgName)
+{
+  int nFlds = msg.mimeFieldsCount();
+
+  logFile << std::endl << msgName << ':' << std::endl;
+
+  if (nFlds) {
+    MimeField iterFld(msg, 0);
+
+    for (int i = 0; i < nFlds; ++i) {
+      MimeField currFld(msg, i);
+
+      ALWAYS_ASSERT(sameMimeField(currFld, iterFld));
+      iterFld = iterFld.next();
+
+      MimeField fFld(msg, currFld.nameGet());
+      ALWAYS_ASSERT(fFld.valid());
+
+      dumpMimeField(currFld);
+    }
+    ALWAYS_ASSERT(!iterFld.valid());
+  }
+}
+
+void
+doCrap(MsgBase &msg, bool add)
+{
+  MimeField f(msg, "x-crap");
+
+  ALWAYS_ASSERT(f.valuesCount() == 3);
+  ALWAYS_ASSERT(f.valGet(0) == "one");
+  ALWAYS_ASSERT(f.valGet(1) == "two");
+  ALWAYS_ASSERT(f.valGet(2) == "three");
+
+  MimeField fd(f.nextDup());
+
+  ALWAYS_ASSERT(fd.valuesCount() == 1);
+  ALWAYS_ASSERT(fd.valGet(0) == "four");
+
+  if (add) {
```

Comment:
In my test environment, this code to change the  edit the values of a field works when '--enable-debug' is used on configure, but fails when it is not used.